### PR TITLE
switch train validation to use share()'d agent, switch agents to share model always

### PR DESCRIPTION
**Pull request by alexholdenmiller**
separate train vs valid agent by using a shared agent during valid. this means shared instances must be ready to predict.

- migrate seq2seq, ibm_seq2seq, fairseq, memnn, language_model, drqa to new setup

### CODE REVIEW

File: parlai/scripts/train_model.py [hunk 1]
```diff
@@ -255,7 +257,13 @@ def log(self):
...
         # time elapsed
         logs.append('time:{}s'.format(math.floor(self.train_time.time())))
-        logs.append('total_exs:{}'.format(self.world.get_total_exs()))
+        total_exs = self.world.get_total_exs()
+        logs.append('total_exs:{}'.format(total_exs))
+
+        exs_per_ep = self.world.num_examples()
+        if exs_per_ep:
+            logs.append('total_epochs:{}'.format(
+                round(total_exs / exs_per_ep, 2)))
```
Comment (alexholdenmiller) [inline parlai/scripts/train_model.py]: oops this is extra and should be merged already in a slightly modified form, ignore this section

**Comment by alexholdenmiller** [comment]
this will need us to start versioning ParlAI

**Comment by stephenroller** [comment]
Eyeballing fairseq lgtm. Didn't actually try to run it, but happy to try it if desired.

File: parlai/scripts/train_model.py [hunk 2]
```diff
@@ -103,12 +103,14 @@ def run_eval(agent, opt, datatype, max_exs=-1, write_log=False, valid_world=None
...
-    if opt.get('evaltask'):
-        opt['task'] = opt['evaltask']
 
     if valid_world is None:
-        valid_world = create_task(opt, agent)
+        opt = opt.copy()
+        opt['datatype'] = datatype
+        if opt.get('evaltask'):
+            opt['task'] = opt['evaltask']
+        valid_agent = create_agent_from_shared(agent.share())
```
Comment (jaseweston) [inline parlai/scripts/train_model.py]: hmm can we make this optional perhaps in case for models that it doesnt work for?

Comment (alexholdenmiller) [reply parlai/scripts/train_model.py]: I'll put this behind a (default False) flag for now

**Comment by alexholdenmiller** [comment]
going to take your previous "approve" as law and merge this

**Pull request merged**